model_signing/hashing/model_hash.py

```diff
@@ -0,0 +1,92 @@
+# Copyright (c) 2024, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""This package provides functionality to hash the model folder and create a statement."""
+from typing import Optional
+
+import hashlib
+import os
+import pathlib
+
+from concurrent import futures
+
+from in_toto_attestation.v1 import statement 
+from in_toto_attestation.v1 import resource_descriptor as res_desc
+
+
+_CHUNK_SIZE = int(1_000_000_000) # ~1GB
+_MAX_WORKERS = 10
+
+
+def __hash_file(path: pathlib.Path) -> bytes:
+    h = hashlib.sha256()
+    with open(path, 'rb') as fd:
+        while True:
+            data = fd.read(_CHUNK_SIZE)
+            if not data:
+                break
+            h.update(data)
+    return h.hexdigest()
+
+
+def __create_subject(full_path: pathlib.Path, internal_path: str) -> res_desc.ResourceDescriptor:
+    return res_desc.ResourceDescriptor(
+        # TODO: clean path; make sure to only store parts that are in the model folder.
+        # e.g. instead of /a/b/c/model_base/weights/chunk1-1000.gz only store weights/chunk1-100.gz
+        name=internal_path,
+        digest={'sha256': __hash_file(full_path)}
+    )
+
+
+def hash_model(
+        path: str,
+        ignore: Optional[list[str]] = ['signature.json'],
+        model_metadata: dict[str, str]={'name': 'unknown'}) -> statement.Statement:
```

Comment:
was curious if e should make this function return bytes or something generic, so that if we ever change the implementation, it won't be a breaking change.